### PR TITLE
manifest: support interface checking

### PR DESCRIPTION
**Pull request by fyrchik**
There are some standards (NEP5, etc.) which impose
some restrictions on what methods and events a contract
must contain and their signatures. This commit supports
checking if arbitrary manifest complies with the standard.

### CODE REVIEW

**Comment by roman-khimov** [comment]
Let's move compiler part out of this PR. We can't really change manifest checks now because neo-project/neo#1883 is not (yet) accepted (we can make it an extension, but it's not critical for us).

**Comment by roman-khimov** [comment]
OK, I see you're not changing contract deployment checks now, so this makes more sense as it is now.

File: cli/smartcontract/smart_contract.go [hunk 1]
```diff
@@ -150,6 +150,10 @@ func NewCommands() []cli.Command {
 						Name:  "config, c",
 						Usage: "Configuration input file (*.yml)",
 					},
+					cli.BoolFlag{
+						Name:  "check-standards",
```
Comment (roman-khimov) [inline cli/smartcontract/smart_contract.go:154]: I think it should be the default behavior, this switch can turn it off if needed.

File: cli/contract_test.go [hunk 2]
```diff
@@ -104,3 +104,55 @@ func TestComlileAndInvokeFunction(t *testing.T) {
 		require.Equal(t, []byte("on update|sub update"), res.Stack[0].Value())
 	})
 }
+
+func TestCompileExamples(t *testing.T) {
```
Comment (roman-khimov) [inline cli/contract_test.go:108]: But we already have something like that in the compiler.

File: cli/smartcontract/smart_contract.go [hunk 3]
```diff
@@ -154,6 +154,10 @@ func NewCommands() []cli.Command {
 						Name:  "check-standards",
 						Usage: "check supported standards compliance",
 					},
+					cli.BoolFlag{
+						Name:  "check-events",
```
Comment (roman-khimov) [inline cli/smartcontract/smart_contract.go:158]: It also can be the default.

File: pkg/compiler/codegen.go [hunk 4]
```diff
@@ -870,6 +873,10 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
 			ast.Walk(c, n.Fun)
 			emit.Opcodes(c.prog.BinWriter, opcode.CALLA)
 		case isSyscall(f):
+			if f.name == "Notify" {
```
Comment (roman-khimov) [inline pkg/compiler/codegen.go:876]: Check package also?

File: pkg/compiler/codegen.go [hunk 5]
```diff
@@ -870,6 +873,10 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
...
+				tv := c.typeAndValueOf(n.Args[0])
+				c.emittedEvents[tv.Value.String()] = struct{}{}
```
Comment (roman-khimov) [inline pkg/compiler/codegen.go:878]: This one is weak in that only names are checked.

**Review by roman-khimov — CHANGES_REQUESTED** [review]
Please fix the linter also.

File: pkg/compiler/codegen.go (hunk 4 above)
Comment (fyrchik) [reply pkg/compiler/codegen.go:876]: There is a single syscall with that name, but yeah, better check a package also.

File: cli/contract_test.go (hunk 2 above)
Comment (fyrchik) [reply cli/contract_test.go:108]: It is here mainly to cover compiler options settings. Maybe we can cut this test to just 3 examples (both tokens and contract with invalid events).

File: cli/smartcontract/smart_contract.go (hunk 3 above)
Comment (fyrchik) [reply cli/smartcontract/smart_contract.go]: Fixed.

File: pkg/compiler/codegen.go (hunk 5 above)
Comment (fyrchik) [reply pkg/compiler/codegen.go]: Fixed.

File: cli/smartcontract/smart_contract.go (hunk 1 above)
Comment (fyrchik) [reply cli/smartcontract/smart_contract.go]: Fixed.

File: pkg/compiler/codegen.go [hunk 6]
```diff
@@ -870,6 +873,15 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
 			ast.Walk(c, n.Fun)
 			emit.Opcodes(c.prog.BinWriter, opcode.CALLA)
 		case isSyscall(f):
+			if f.pkg.Name() == "runtime" && f.name == "Notify" {
+				tv := c.typeAndValueOf(n.Args[0])
+				params := make([]string, 0, len(n.Args[1:]))
+				for _, p := range n.Args[1:] {
+					params = append(params, c.scTypeFromExpr(p))
```
Comment (roman-khimov) [inline pkg/compiler/codegen.go:880]: It's a little risky, but let's try doing that.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**